File: sdk/typescript/src/adapters/llamaindex.ts
import { TraceLens } from '../tracer';
import { Span } from '../span';

export class TraceLensLlamaIndexHandler {
  private tracer: TraceLens;
  private activeSpans: Map<string, { span: Span; traceId: string }> = new Map();
  private defaultTraceId: string;

  constructor(options: { tracer: TraceLens; traceId?: string }) {
    this.tracer = options.tracer;
    this.defaultTraceId = options.traceId || `llamaindex-${Date.now()}`;
  }

  onRetrieveStart(query: string, eventId: string): void {
    const span = this.tracer.startSpan({
      traceId: this.defaultTraceId,
      agent: 'llamaindex',
      task: 'vector_retrieval',
    });
    span.setAttribute('query', query);
    span.setAttribute('framework', 'llamaindex');
    this.activeSpans.set(eventId, { span, traceId: this.defaultTraceId });
  }

  onRetrieveEnd(nodes: any[], eventId: string): void {
    const item = this.activeSpans.get(eventId);
    if (!item) return;

    item.span.setAttribute('retrieved_chunks_count', nodes?.length || 0);
    item.span.end({ status: 'ok' });
    this.activeSpans.delete(eventId);
  }

  onQueryStart(query: string, eventId: string): void {
    const span = this.tracer.startSpan({
      traceId: this.defaultTraceId,
      agent: 'llamaindex',
      task: 'rag_query_execution',
    });
    span.setAttribute('query', query);
    span.setAttribute('framework', 'llamaindex');
    this.activeSpans.set(eventId, { span, traceId: this.defaultTraceId });
  }

  onQueryEnd(response: any, eventId: string): void {
    const item = this.activeSpans.get(eventId);
    if (!item) return;

    item.span.setAttribute('response', String(response).slice(0, 1000));
    item.span.end({ status: 'ok' });
    this.activeSpans.delete(eventId);
  }
}

File: sdk/python/tracelens/adapters/llamaindex.py
import time
from typing import Any, Dict, List, Optional
from tracelens.tracer import TraceLens
from tracelens.span import Span

class TraceLensLlamaIndexHandler:
    """
    LlamaIndex Callback Handler for TraceLens.
    Captures vector search, chunk retrieval, and RAG query execution.
    """

    def __init__(self, tracer: TraceLens, trace_id: Optional[str] = None):
        self.tracer = tracer
        self.default_trace_id = trace_id or f"llamaindex-{int(time.time() * 1000)}"
        self.active_events: Dict[str, Span] = {}

    def on_retrieve_start(self, query: str, event_id: str) -> None:
        span = self.tracer.start_span(
            trace_id=self.default_trace_id,
            agent="llamaindex",
            task="vector_retrieval",
        )
        span.set_attribute("query", query)
        span.set_attribute("framework", "llamaindex")
        self.active_events[event_id] = span

    def on_retrieve_end(self, nodes: List[Any], event_id: str) -> None:
        span = self.active_events.pop(event_id, None)
        if not span:
            return

        span.set_attribute("retrieved_chunks_count", len(nodes) if nodes else 0)
        span.end(status="ok")

    def on_query_start(self, query: str, event_id: str) -> None:
        span = self.tracer.start_span(
            trace_id=self.default_trace_id,
            agent="llamaindex",
            task="rag_query_execution",
        )
        span.set_attribute("query", query)
        span.set_attribute("framework", "llamaindex")
        self.active_events[event_id] = span

    def on_query_end(self, response: Any, event_id: str) -> None:
        span = self.active_events.pop(event_id, None)
        if not span:
            return

        span.set_attribute("response", str(response)[:1000])
        span.end(status="ok")
